lib/PrintfPass.cpp

```diff
@@ -0,0 +1,318 @@
+// Copyright 2022 The Clspv Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "clspv/Option.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+
+#include "Constants.h"
+#include "PushConstant.h"
+
+#include "PrintfPass.h"
+
+using namespace llvm;
+
+std::string clspv::PrintfPass::GetStringLiteral(Value *Val) {
+  GlobalVariable *GlobalVal = nullptr;
+  Instruction *ConstExprEquivInstr = nullptr;
+  // If opaque pointers aren't enabled, the global may be hidden behind a
+  // GEP constexpr
+  if (auto *CE = dyn_cast<ConstantExpr>(Val)) {
+    ConstExprEquivInstr = CE->getAsInstruction();
+    if (auto *GEPInstr = dyn_cast<GetElementPtrInst>(ConstExprEquivInstr)) {
+      auto *Ptr = GEPInstr->getPointerOperand();
+      if (auto *GV = dyn_cast<GlobalVariable>(Ptr)) {
+        GlobalVal = GV;
+      }
+    }
+  }
+
+  // If there wasn't a GEP then the global will be used directly instead
+  if (!GlobalVal) {
+    GlobalVal = dyn_cast<GlobalVariable>(Val);
+  }
+
+  if (GlobalVal) {
+    auto *Initializer = GlobalVal->getInitializer();
+    if (auto *InitArray = dyn_cast<ConstantDataArray>(Initializer)) {
+      if (InitArray->isCString()) {
+        if (ConstExprEquivInstr)
+          ConstExprEquivInstr->deleteValue();
+        return InitArray->getAsCString().str();
+      }
+    }
+  }
+
+  return "";
+}
+
+unsigned clspv::PrintfPass::GetPrintfStoreSize(const DataLayout &DL, Type *Ty) {
+  return std::max(4ul, DL.getTypeStoreSize(Ty).getFixedValue());
```

Comment:
```suggestion
  return std::max(4ul, (unsigned long)DL.getTypeStoreSize(Ty).getFixedValue());
```

Reply:
It seems that on windows and macos `getFixedValue` returns an `unsigned long long`, so it needs to be casted to avoid a type mismatched in `std::max`.